core/src/main/java/hudson/tasks/ArtifactArchiver.java

```diff
@@ -214,20 +214,10 @@ public boolean isCaseSensitive() {
         this.caseSensitive = caseSensitive;
     }
 
-    private void listenerWarnOrError(TaskListener listener, String message) {
-    	if (allowEmptyArchive) {
-    		listener.getLogger().println(String.format("WARN: %s", message));
-    	} else {
-    		listener.error(message);
-    	}
-    }
-
     @Override
-    public void perform(Run<?,?> build, FilePath ws, Launcher launcher, TaskListener listener) throws InterruptedException, AbortException {
+    public void perform(Run<?,?> build, FilePath ws, Launcher launcher, TaskListener listener) throws IOException, InterruptedException {
         if(artifacts.length()==0) {
-            listener.error(Messages.ArtifactArchiver_NoIncludes());
-            build.setResult(Result.FAILURE);
-            return;
+            throw new AbortException(Messages.ArtifactArchiver_NoIncludes());
```

Comment:
Does this really not change the semantics for freestyle projects? E.g. other publishers will still run?

Reply:
AFAICT `AbstractBuildExecution.performAllBuildSteps` treats an `AbortException` much like a `return false`.